File: app/src/main/java/de/westnordost/osmagent/quests/osm/download/ElementGeometryCreator.java
package de.westnordost.osmagent.quests.osm.download;

import com.google.common.collect.Lists;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.ListIterator;

import de.westnordost.osmagent.meta.OsmAreas;
import de.westnordost.osmagent.quests.osm.ElementGeometry;
import de.westnordost.osmapi.map.data.Element;
import de.westnordost.osmapi.map.data.LatLon;
import de.westnordost.osmapi.map.data.Node;
import de.westnordost.osmapi.map.data.Relation;
import de.westnordost.osmapi.map.data.RelationMember;
import de.westnordost.osmapi.map.data.Way;

public class ElementGeometryCreator
{
	private GeometryMapDataProvider data;

	public ElementGeometryCreator(GeometryMapDataProvider data)
	{
		this.data = data;
	}

	public ElementGeometry create(Node node)
	{
		return new ElementGeometry(node.getPosition());
	}

	public ElementGeometry create(Way way)
	{
		List<LatLon> polyline = createWayGeometry(way.getNodeIds());
		// unable to create geometry
		if(polyline.isEmpty()) return null;

		List<List<LatLon>> polylines = new ArrayList<>(1);
		polylines.add(polyline);

		if(OsmAreas.isArea(way))
		{
			return new ElementGeometry(null, polylines);
		}
		else
		{
			return new ElementGeometry(polylines, null);
		}
	}

	public ElementGeometry create(Relation relation)
	{
		if(OsmAreas.isArea(relation))
		{
			return createMultipolygonGeometry(relation);
		}
		else
		{
			return createPolylinesGeometry(relation);
		}
	}

	private List<LatLon> createWayGeometry(List<Long> wayNodeIds)
	{
		List<LatLon> geometry = new ArrayList<>(wayNodeIds.size());
		for(long nodeId : wayNodeIds)
		{
			geometry.add(data.getNode(nodeId).getPosition());
		}
		return geometry;
	}

	private List<List<Long>> getWaysOfRelationWithRole(Relation relation, String role)
	{
		List<List<Long>> result = new ArrayList<>();
		for(RelationMember member : relation.getMembers())
		{
			if(member.getType() != Element.Type.WAY) continue;

			long wayId = member.getRef();
			if(role == null || role.equals(member.getRole()))
			{
				List<Long> nodeIds = data.getWay(wayId).getNodeIds();
				if(nodeIds.size() > 1)
				{
					result.add(nodeIds);
				}
			}
		}
		return result;
	}

	private ElementGeometry createPolylinesGeometry(Relation relation)
	{
		List<List<Long>> waysNodeIds = getWaysOfRelationWithRole(relation, null);
		ConnectedWays ways = joinWays(waysNodeIds);

		List<List<Long>> joinedWaysNodeIds = ways.rest;
		joinedWaysNodeIds.addAll(ways.rings);

		List<List<LatLon>> polylines = createWaysGeometry(joinedWaysNodeIds);

		// no valid geometry
		if(polylines.isEmpty()) return null;

		return new ElementGeometry(polylines, null);
	}

	private ElementGeometry createMultipolygonGeometry(Relation relation)
	{
		List<List<LatLon>> rings = new ArrayList<>();
		rings.addAll(createNormalizedRingGeometry(relation, "outer", false));
		rings.addAll(createNormalizedRingGeometry(relation, "inner", true));
		// no valid geometry
		if(rings.isEmpty()) return null;

		return new ElementGeometry(null, rings);
	}

	private List<List<LatLon>> createNormalizedRingGeometry(Relation relation, String role,
															boolean clockwise)
	{
		List<List<Long>> waysNodeIds = getWaysOfRelationWithRole(relation, role);
		List<List<Long>> ringsNodeIds = joinWays(waysNodeIds).rings;
		List<List<LatLon>> ringGeometry = createWaysGeometry(ringsNodeIds);
		setOrientation(ringGeometry, clockwise);
		return ringGeometry;
	}

	private List<List<LatLon>> createWaysGeometry(List<List<Long>> waysNodeIds)
	{
		List<List<LatLon>> result = new ArrayList<>(waysNodeIds.size());
		for(List<Long> wayNodeIds : waysNodeIds)
		{
			result.add(createWayGeometry(wayNodeIds));
		}
		return result;
	}

	/** Ensures that all given rings are defined in clockwise/counter-clockwise direction */
	private static void setOrientation(List<List<LatLon>> rings, boolean clockwise)
	{
		for(List<LatLon> ring : rings)
		{
			if(isRingDefinedClockwise(ring) != clockwise)
			{
				Collections.reverse(ring);
			}
		}
	}

	private static class ConnectedWays
	{
		List<List<Long>> rings = new ArrayList<>();
		List<List<Long>> rest = new ArrayList<>();
	}

	private static ConnectedWays joinWays(List<List<Long>> waysNodeIds)
	{
		NodeWayMap nodeWayMap = new NodeWayMap(waysNodeIds);

		ConnectedWays result = new ConnectedWays();

		List<Long> currentWay = new ArrayList<>();

		while(nodeWayMap.hasNextNodeId())
		{
			Long nodeId;
			if(currentWay.isEmpty())
			{
				nodeId = nodeWayMap.getNextNodeId();
			}
			else
			{
				nodeId = currentWay.get(currentWay.size()-1);
			}

			List<List<Long>> waysAtNode = nodeWayMap.getWaysAtNode(nodeId);
			if(waysAtNode == null)
			{
				result.rest.add(currentWay);
				currentWay = new ArrayList<>();
			}
			else
			{
				List<Long> way = waysAtNode.get(0);

				addTo(way, currentWay);
				nodeWayMap.removeWay(way);

				// finish ring and start new one
				if(isRing(currentWay))
				{
					result.rings.add(currentWay);
					currentWay = new ArrayList<>();
				}
			}
		}
		if(!currentWay.isEmpty())
		{
			result.rest.add(currentWay);
		}

		return result;
	}

	private static boolean isRing(List<Long> way)
	{
		return way.get(0).equals(way.get(way.size() - 1));
	}

	/** add <tt>way</tt> to the end of <tt>polyWay</tt>, if necessary in reverse */
	private static void addTo(List<Long> way, List<Long> polyWay)
	{
		if(polyWay.isEmpty())
		{
			polyWay.addAll(way);
		}
		else
		{
			long addLast = way.get(way.size() - 1);
			long toLast = polyWay.get(polyWay.size() - 1);
			if(addLast == toLast)
			{
				way = Lists.reverse(way);
			}
			// +1 to not add the first vertex because it has already been added
			ListIterator<Long> it = way.listIterator(1);
			while(it.hasNext())
			{
				polyWay.add(it.next());
			}
		}
	}

	private static boolean isRingDefinedClockwise(List<LatLon> ring)
	{
		double sum = 0;
		for(int i=0; i<ring.size(); ++i)
		{
			LatLon pos1 = ring.get(i);
			LatLon pos2 = ring.get((i+1) % ring.size());
			sum += (pos2.getLongitude() - pos1.getLongitude()) * (pos2.getLatitude() + pos1.getLatitude());
		}
		return sum > 0;
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/data/osm/elementgeometry/ElementGeometryCreator.kt
package de.westnordost.streetcomplete.data.osm.elementgeometry

import de.westnordost.osmapi.map.data.*
import de.westnordost.streetcomplete.ktx.isArea
import de.westnordost.streetcomplete.util.centerPointOfPolygon
import de.westnordost.streetcomplete.util.centerPointOfPolyline
import de.westnordost.streetcomplete.util.isRingDefinedClockwise
import javax.inject.Inject
import kotlin.collections.ArrayList

/** Creates an ElementGeometry from an element and a collection of positions. */
class ElementGeometryCreator @Inject constructor() {

    /** Create an ElementPointGeometry for a node. */
    fun create(node: Node) = ElementPointGeometry(node.position)

    /**
     * Create an ElementGeometry for a way
     *
     * @param way the way to create the geometry for
     * @param wayGeometry the geometry of the way: A list of positions of its nodes.
     *
     * @return an ElementPolygonsGeometry if the way is an area or an ElementPolylinesGeometry
     *          if the way is a linear feature */
    fun create(way: Way, wayGeometry: List<LatLon>): ElementGeometry? {
        val polyline = ArrayList(wayGeometry)
        polyline.eliminateDuplicates()
        if (wayGeometry.size < 2) return null

        return if (way.isArea()) {
            /* ElementGeometry considers polygons that are defined clockwise holes, so ensure that
               it is defined CCW here. */
            if (polyline.isRingDefinedClockwise()) polyline.reverse()
            ElementPolygonsGeometry(arrayListOf(polyline), polyline.centerPointOfPolygon())
        } else {
            ElementPolylinesGeometry(arrayListOf(polyline), polyline.centerPointOfPolyline())
        }
    }

    /**
     * Create an ElementGeometry for a relation
     *
     * @param relation the relation to create the geometry for
     * @param wayGeometries the geometries of the ways that are members of the relation. It is a
     *                      map of way ids to a list of positions.
     *
     * @return an ElementPolygonsGeometry if the relation describes an area or an
     *         ElementPolylinesGeometry if it describes is a linear feature */
    fun create(relation: Relation, wayGeometries: Map<Long, List<LatLon>>): ElementGeometry? {
        return if (relation.isArea()) {
            createMultipolygonGeometry(relation, wayGeometries)
        } else {
            createPolylinesGeometry(relation, wayGeometries)
        }
    }

    private fun createMultipolygonGeometry(
        relation: Relation,
        wayGeometries: Map<Long, List<LatLon>>
    ): ElementPolygonsGeometry? {

        val outer = createNormalizedRingGeometry(relation, "outer", false, wayGeometries)
        val inner = createNormalizedRingGeometry(relation, "inner", true, wayGeometries)
        if (outer.isEmpty()) return null

        val rings = ArrayList<ArrayList<LatLon>>()
        rings.addAll(outer)
        rings.addAll(inner)

        /* only use first ring that is not a hole if there are multiple
           this is the same behavior as Leaflet or Tangram */
        return ElementPolygonsGeometry(rings, outer.first().centerPointOfPolygon())
    }

    private fun createPolylinesGeometry(
        relation: Relation,
        wayGeometries: Map<Long, List<LatLon>>
    ): ElementPolylinesGeometry? {

        val waysNodePositions = getRelationMemberWaysNodePositions(relation, wayGeometries)
        val joined = waysNodePositions.joined()

        val polylines = joined.ways
        polylines.addAll(joined.rings)
        if (polylines.isEmpty()) return null

        /* if there are more than one polylines, these polylines are not connected to each other,
           so there is no way to find a reasonable "center point". In most cases however, there
           is only one polyline, so let's just take the first one...
           This is the same behavior as Leaflet or Tangram */
        return ElementPolylinesGeometry(polylines, polylines.first().centerPointOfPolyline())
    }

    private fun createNormalizedRingGeometry(
        relation: Relation,
        role: String,
        clockwise: Boolean,
        wayGeometries: Map<Long, List<LatLon>>
    ): ArrayList<ArrayList<LatLon>> {

        val waysNodePositions = getRelationMemberWaysNodePositions(relation, role, wayGeometries)
        val ringGeometry = waysNodePositions.joined().rings
        ringGeometry.setOrientation(clockwise)
        return ringGeometry
    }

    private fun getRelationMemberWaysNodePositions(
        relation: Relation, wayGeometries: Map<Long, List<LatLon>>
    ): List<List<LatLon>> {
        return relation.members.filter { it.type == Element.Type.WAY }.mapNotNull {
            getValidNodePositions(wayGeometries[it.ref])
        }
    }

    private fun getRelationMemberWaysNodePositions(
        relation: Relation, withRole: String, wayGeometries: Map<Long, List<LatLon>>
    ): List<List<LatLon>> {
        return relation.members.filter { it.type == Element.Type.WAY && it.role == withRole }.mapNotNull {
            getValidNodePositions(wayGeometries[it.ref])
        }
    }

    private fun getValidNodePositions(wayGeometry: List<LatLon>?): List<LatLon>? {
        if (wayGeometry == null) return null
        val nodePositions = ArrayList(wayGeometry)
        nodePositions.eliminateDuplicates()
        return if (nodePositions.size >= 2) nodePositions else null
    }
}

/** Ensures that all given rings are defined in clockwise/counter-clockwise direction  */
private fun List<MutableList<LatLon>>.setOrientation(clockwise: Boolean) {
    for (ring in this) {
        if (ring.isRingDefinedClockwise() != clockwise) {
            ring.reverse()
        }
    }
}

private fun List<LatLon>.isRing() = first() == last()

private class ConnectedWays(val rings: ArrayList<ArrayList<LatLon>>, val ways: ArrayList<ArrayList<LatLon>>)

/** Returns a list of polylines joined together at their endpoints into rings and ways */
private fun List<List<LatLon>>.joined(): ConnectedWays {
    val nodeWayMap = NodeWayMap(this)

    val rings: ArrayList<ArrayList<LatLon>> = ArrayList()
    val ways: ArrayList<ArrayList<LatLon>> = ArrayList()

    var currentWay: ArrayList<LatLon> = ArrayList()

    while (nodeWayMap.hasNextNode()) {
        val node: LatLon = if (currentWay.isEmpty()) nodeWayMap.getNextNode() else currentWay.last()

        val waysAtNode = nodeWayMap.getWaysAtNode(node)
        if (waysAtNode == null) {
            ways.add(currentWay)
            currentWay = ArrayList()
        } else {
            val way = waysAtNode.first()

            currentWay.join(way)
            nodeWayMap.removeWay(way)

            // finish ring and start new one
            if (currentWay.isRing()) {
                rings.add(currentWay)
                currentWay = ArrayList()
            }
        }
    }
    if (currentWay.isNotEmpty()) {
        ways.add(currentWay)
    }

    return ConnectedWays(rings, ways)
}

/** Join the given adjacent polyline into this polyline */
private fun MutableList<LatLon>.join(way: List<LatLon>) {
    if (isEmpty()) {
        addAll(way)
    } else {
        when {
            last() == way.last() -> addAll(way.asReversed().subList(1, way.size))
            last() == way.first() -> addAll(way.subList(1, way.size))
            first() == way.last() -> addAll(0, way.subList(0, way.size - 1))
            first() == way.first() -> addAll(0, way.asReversed().subList(0, way.size - 1))
            else -> throw IllegalArgumentException("The ways are not adjacent")
        }
    }
}

private fun MutableList<LatLon>.eliminateDuplicates() {
    val it = iterator()
    var prev: LatLon? = null
    while (it.hasNext()) {
        val line = it.next()
        if (prev == null || line.latitude != prev.latitude || line.longitude != prev.longitude) {
            prev = line
        } else {
            it.remove()
        }
    }
}
